### PR TITLE
Add thread to remove old measurements

### PR DESCRIPTION
**Pull request by Mast3rwaf1z**
When the localizationService is created, a thread is also started which runs while the object contains data.

The data is also now stored in an object MeasurementGroup, which contains the list of measurements and a timestamp for each measurement.

The following command can be used  to test it:
```sh
while true; do 
    curl -X 'POST' -H 'content-type: application/json' -d '{"trackerId":1, "measurements":[{"type":1, "macAddress":"1", "value":"1"}]}' 'http://localhost:8081/localization/measurements/add' 
    sleep 30
done
```
after 2 minutes, there should be at most 4 measurements in the measurementgroup

There is a print statement that should be removed once the PR is approved, used for testing

### CODE REVIEW

**Comment by Mast3rwaf1z** [comment]
I want feedback and maybe improvements on how the measurementGroup class is implemented and how it could be implemented better

**Review by Casper-NS — APPROVED** [review]

File: CentralHub.Api/Services/LocalizationService.cs [hunk 1]
```diff
@@ -4,20 +4,63 @@ namespace CentralHub.Api.Services;
...
+                    {
+                        _measurements[key].Remove(group);
+                    }
+                }
+                if (_measurements.Keys.All(key => _measurements[key].Count == 0))
+                {
+                    return;
+                }
             }
+            Thread.Sleep(delay * 1000);
```
Comment (mads256h) [inline CentralHub.Api/Services/LocalizationService.cs:63]: Sleep until the next group needs to be removed

**Review by mads256h — CHANGES_REQUESTED** [review]

File: CentralHub.Api/Services/LocalizationService.cs (hunk 1 above)
Comment (Mast3rwaf1z) [reply CentralHub.Api/Services/LocalizationService.cs:63]: that is what it does? it calculates delay during the thread's execution

Comment (Mast3rwaf1z) [reply CentralHub.Api/Services/LocalizationService.cs:63]: i should update it to use milliseconds in the calculation instead of multiplying by 1000 tho

Comment (mads256h) [reply CentralHub.Api/Services/LocalizationService.cs]: Ok I see it just seems strange to me that delay is initialized to 120

File: CentralHub.Api/Services/LocalizationService.cs [hunk 2]
```diff
@@ -4,20 +4,63 @@ namespace CentralHub.Api.Services;
...
-                value.AddRange(measurements);
+                value.Add(new MeasurementGroup(measurements.ToList()));
             }
             else
             {
-                _measurements.Add(id, measurements.ToList());
+                _measurements.Add(id, new List<MeasurementGroup>() { new MeasurementGroup(measurements.ToList()) });
+            }
+        }
+        if (MeasurementRemover == null || !validStates.Contains(MeasurementRemover.ThreadState))
```
[the remark this answers is not in the document]
Comment (mads256h) [reply CentralHub.Api/Services/LocalizationService.cs:23]: Nevermind

File: CentralHub.Api/Services/LocalizationService.cs [hunk 3]
```diff
@@ -4,20 +4,63 @@ namespace CentralHub.Api.Services;
...
+        {
+            MeasurementRemover = new Thread(new ThreadStart(RemoveMeasurements));
```
Comment (mads256h) [inline CentralHub.Api/Services/LocalizationService.cs:25]: I think we should join the old thread here if there is any

**Review by mads256h — CHANGES_REQUESTED** [review]

File: CentralHub.Api/Services/LocalizationService.cs (hunk 1 above)
Comment (Mast3rwaf1z) [reply CentralHub.Api/Services/LocalizationService.cs]: Max value

Comment (mads256h) [reply CentralHub.Api/Services/LocalizationService.cs]: Set it to int.MaxValue

Comment (mads256h) [reply CentralHub.Api/Services/LocalizationService.cs]: There is no need to wakeup to do nothing

**Review by mads256h — APPROVED** [review]
LGTM

**Comment by mads256h** [comment]
Merged in another branch

**Pull request closed without merging**